### PR TITLE
Copy to sequence stages + optimizer + s3 oneshot sink stub

### PR DESCRIPTION
**Pull request by moulimukherjee**
These are the next set of changes after https://github.com/MaterializeInc/materialize/pull/24355 which implements the optimizer for COPY TO statements and the initial plumbing to start the s3 sink. The s3 sink implementation will be done in a later PR.

### Motivation
This PR implements the optimizer for COPY TO statements and the initial plumbing to start the s3 sink as part of https://github.com/MaterializeInc/materialize/issues/7256

### Tips for reviewer

Inline + comments

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/sequencer/inner/copy_to.rs
Comment (moulimukherjee) [inline src/adapter/src/coord/sequencer/inner/copy_to.rs:731]: This is copied from peeks. I am not exactly sure what this does or if this will be needed for COPY TO

**Comment by mjibson** [comment]
I think that you should be able to co-opt the existing peek pipeline instead of copy-pasting all this, which will bitrot quickly. (For example, I merged a PR earlier today that refactors how the optimize lir step works.) The explain code already does some work to do a similar thing here where explain and select want to share almost all of their paths, but not quite. COPY TO should be able to do a similar thing. Various parts of the PeekStage inner structs have a `explain_ctx` field. Change that to be an enum with Peek, Explain, and CopyTo variants. (Its current None variant will be the new Peek variant.) That should allow the copy to code to branch off near the end but share the rest of the pipeline.

**Comment by mjibson** [comment]
Strongly recommend if that's a good idea that you make a new PR that only does the `explain_ctx: Option<ExplainContext>` -> `output: PeekOutput` refactor with only the Peek and Explain variants. Then adding copy to to it in another PR will be much easier to follow.

**Comment by moulimukherjee** [comment]
I agree with the potential bitrot. I had actually tried that approach too, but gave it up because of constant branching, like there's no fast path, sinks needed a realtion desc plumbed through (not available to select). Let me give it a try again though, I was using existing `copy_to` in peek to pass extra things. Having it part of `explain_ctx` might make it better.

**Comment by moulimukherjee** [comment]
Updated the implementation to use existing peek path with additional copy context (code's still rough around the edges, have not cleaned up yet). Able to start the sink dataflow. There's no plumbing yet to send back a proper response.
Will create a separate PR to refactor `explain_ctx: Option<ExplainContext>` to an enum for Peek and Explain and get that merged first.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [inline src/adapter/src/coord/sequencer/inner/peek.rs:724]: I would have expected to see the CopyTo variant somewhere here?

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [inline src/adapter/src/coord/sequencer/inner/peek.rs:756]: I would expect to see the CopyTo variant here as like a PeekStage::CopyTo. From my read, the stuff in peek_stage_finish is not needed by copy?

File: src/adapter/src/optimize/peek.rs
Comment (mjibson) [inline src/adapter/src/optimize/peek.rs:328]: To me this seems like it belongs somewhere else (not sure where tho)? @aalexandrov do you have thoughts? (I don't know much about stages MIR and below though, so this could be correct). Maybe someone from storage also has a thought.

**Review by mjibson — COMMENTED** [review]
This looks wayyy better to me! I support this approach.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs:756]: `peek_stage_finish` is what finally calls the `implement...` method to schedule the dataflow. Since even copy to needs a dataflow, I forked the logic there instead of creating a different stage. It's also emitting optimizer notices, doing some `install_watch_set` calls which I thought might be needed for the copy to as well?

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:328]: So, this part is doing a HirScalar -> MirScalar -> String -> Uri for the copy target before passing that value to the s3 sink. Since the `LocalMirPlan` is already supposed to have Mir, maybe the Hir -> Mir could have happened earlier. But the validation upon the s3 url itself can only be done after getting the `Uri` value out of it.

Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:328]: Actually let me clean this up a bit. The scalar can be evaluated earlier.

Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:326]: I moved the scalar evaluation earlier. Keeping only the compute sink addition to the dataflow.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs:786]: Separating this out in a different finish stage for copy_to

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [inline src/adapter/src/coord/sequencer/inner/peek.rs:770]: My idea is to have a new variant like `PeekStage::CopyTo` here that will end up calling a similar function as finish, but with only things needed for copy.

File: src/adapter/src/coord.rs
Comment (moulimukherjee) [reply src/adapter/src/coord.rs:391]: ah til

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs]: Just updated

File: src/adapter/src/optimize/peek.rs
Comment (benesch) [inline src/adapter/src/optimize/peek.rs:137]: Just driving by, but you probably don't need these to be different lifetimes. Does this compile if you use `'s` for both? If so, that'll make the code simpler. Though would recommend renaming the lifetime to `'a` rather than `'s` because `'s` looks like it's named for *s*ession and if you're using it for both *s*ession and *p*eek context, better to use a generic name like `'a`.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs]: Updated!

File: src/adapter/src/coord/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/coord/peek.rs]: Yeah, I agree we should not leak resources. @petrosagg would you know if it's possible to have a sink dataflow definition which just logs and returns?

The other option would be to not schedule the dataflow at all for this PR.

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:137]: ah updated, ty

File: src/adapter/src/optimize/peek.rs
Comment (benesch) [inline src/adapter/src/optimize/peek.rs:248]: I think you _should_ be able to again rely on lifetime elision here now that there's only one lifetime in play.

```suggestion
   pub fn resolve(
         self,
         session: &Session,
         stats: Box<dyn StatisticsOracle>,
         peek_ctx: &PeekContext,
     ) -> LocalMirPlan<ResolvedLocal> {
```

File: src/adapter/src/coord/peek.rs
Comment (benesch) [reply src/adapter/src/coord/peek.rs]: Good to verify, but if you're setting both `as_of` and `up_to` on the sink, I think the dataflow should just clean itself up once it realizes it doesn't need to do anything meaningful with the data.

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:248]: I had that originally before doing the two lifetime thing. It's giving the error 
```
missing lifetime specifier
this function's return type contains a borrowed value, but the signature does not say 
whether it is borrowed from `session` or `peek_ctx`
```

Comment (benesch) [reply src/adapter/src/optimize/peek.rs:248]: Ah ok, never mind—this is as simple as it gets then!

**Review by mjibson — APPROVED** [review]
Adapter parts lgtm. Someone else needs to review the compute and optimize code though.

**Comment by teskje** [comment]
The compute rendering changes lgtm (not much going on so far). I just wonder if we are going to run into confusion down the road when we call this an "S3 sink". I think this can cause confusion in two ways:

* In contrast to other compute sinks, this one can only write snapshot data and cannot deal with incremental updates.
* If we add a "proper" S3 (storage) sink in the future, there will be two things in out code called the same way.

Can we find a better name? Maybe `S3OneShot`/`s3_oneshot`?

**Comment by moulimukherjee** [comment]
@teskje Makes sense, changing it. Some other options, `s3_snapshot`, `copy_to_s3`.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (moulimukherjee) [inline src/adapter/src/coord/sequencer/inner/peek.rs:1082]: I have this from the original peek implementation. I am not sure if I need this in COPY TO since there's not going to be a peek call after. Does that sound right?

cc @petrosagg / @aalexandrov

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:352]: cc @petrosagg / @aalexandrov (lmk if I should tag other people)

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [reply src/adapter/src/coord/sequencer/inner/peek.rs:1082]: I think this depends on the dataflow internals. Since the actual export occurs in a dataflow operator, maybe...you don't need this because the dataflow is able to hold back whatever it needs?

Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs:1083]: That's what I thought too (I am fuzzy on how dataflow holds back stuff). Removing this for the time being.

**Comment by teskje** [comment]
> Some other options, `s3_snapshot`, `copy_to_s3`.

I like `s3_snapshot` too! `copy_to_s3` not so much because it mixes in a SQL concept which I'd like to keep the Compute layer separate from.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [reply src/adapter/src/coord/sequencer/inner/peek.rs]: IIRC, this happens with the various `.downgrade` calls on capabilities. Timestamps are held back until explicitly released.

Comment (teskje) [reply src/adapter/src/coord/sequencer/inner/peek.rs]: For a sink it doesn't make sense to try to install a read hold on its output. A sink output cannot be read (at least not in the scope of compute), only an index can. In fact, https://github.com/MaterializeInc/materialize/pull/24487, which I want to merge tomorrow morning will make it an error to try to set a read policy on a non-index compute collection.

File: src/adapter/src/optimize/peek.rs
Comment (teskje) [inline src/adapter/src/optimize/peek.rs:503]: Since a copy-to-S3 is quite different from a peek (no fast path, no index, but a sink). I would expect to see this in a separate optimizer pipeline, e.g. `optimize/copy_to_s3.rs`. Is there a reason that wouldn't work?

Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:503]: That's what I had originally, but it involved a lot of code duplication https://github.com/MaterializeInc/materialize/pull/24562#issuecomment-1915825181. COPY TO is a bit different from peek but it also shares a lot of same logic of spinning up a similar dataflow but with a sink in this case.

File: src/compute/src/sink/s3_oneshot.rs
Comment (moulimukherjee) [inline src/compute/src/sink/s3_oneshot.rs:43]: @petrosagg Is it fine to just log and return `None`? Or what's the bare minimum I should do here so that the dataflow is spun up but is cleaned up as well?

File: src/adapter/src/optimize/peek.rs
Comment (teskje) [inline src/adapter/src/optimize/peek.rs:351]: Should be fine, based on that for SUBSCRIBE we use the same `transient_id` for both the view output and the sink too (i.e. the sink gets `transient_id` as input but is also exported under `transient_id`). Seems to work, although it _is_ a bit weird.

File: src/adapter/src/optimize/peek.rs
Comment (teskje) [reply src/adapter/src/optimize/peek.rs:352]: No, the expression prepping already happens above in this method, so you don't have to do it here again.

**Review by teskje — COMMENTED** [review]
As mentioned in Slack, I'd much prefer keeping the optimizer pipelines for peek and copy-to separate. (Same for the sequencing paths, but that's up to @MaterializeInc/surfaces to decide). I don't want to hold up this PR if it's blocking active work though, so we can defer this to a follow-up.

Some defects I noticed that come from having peek and copy-to optimization intermingled:
* The dataflow name is "oneshot-select" for copy-to dataflows, which has the potential to cause some confusion.
* We are unnecessarily creating an index for the copy-to output data.

The latter is particularly problematic as it can significantly increase the memory usage of the dataflow.

Also, the current version has the defect that copy-to dataflows are never dropped. Replicas that manage to process them up to the `until` frontier will automatically clean up their state, but the compute controller will hold onto some metadata for those dataflows until it receives a `drop_collections` invocation for them. I think we need to handle these dataflows like SUBSCRIBEs in adapter: listen to responses from the compute layer until we hear back that the dataflow has completed, then call `drop_collections` on the compute controller. Obviously that is not possible without the response protocol being implemented, so it can't be done in this PR. Just pointing it out for later.

File: src/adapter/src/coord/peek.rs
Comment (teskje) [reply src/adapter/src/coord/peek.rs]: The dataflow will compact away on the replica, but the `CollectionState` in the compute controller and some data in controller introspection tables will remain until adapter calls `drop_collections` for it.

File: src/compute/src/sink/s3_oneshot.rs
Comment (antiguru) [inline src/compute/src/sink/s3_oneshot.rs:42]: Do you plan on implementing this operator as part of this PR or separately? It seems like it might be a good idea to keep it separate so the size of this PR doesn't increase further, and to make the reviews easier!

Comment (moulimukherjee) [reply src/compute/src/sink/s3_oneshot.rs:42]: Yes, I am planning to do it in a separate PR, this is already a decent size.

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs]: updated to `panic!`

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs]: ty, removed the comment.

**Comment by moulimukherjee** [comment]
> Obviously that is not possible without the response protocol being implemented, so it can't be done in this PR. Just pointing it out for later.

I will do the response protocol and the sink dataflow implementations in later PRs. This PR seems already pretty sizable so going to keep this scoped to only till the sink can be scheduled.

File: src/adapter/src/optimize/peek.rs
Comment (aalexandrov) [reply src/adapter/src/optimize/peek.rs:496]: Having looked at the changes  in `peek::Optimizer`, I have to say I strongly agree with @teskje here. Even if we use the same code at the `Coordinator` level the `Optimizer` level the difference in the optimizer pipeline is sufficient to warrant a different `optimize` module.

I'll try to pick up this branch and come up with something that satisfies both @MaterializeInc/compute and @MaterializeInc/surfaces as a proposal.

**Review by aalexandrov — CHANGES_REQUESTED** [review]
I'm afraid that I cannot accept the proposed changes to `optimize/peek.rs`. 

The difference between the optimization paths of `SELECT` and `COPY TO` is sufficient to for me to warrant a new `optimize/copy_to.rs` module and I won't be comfortable maintaining `optimize/peek.rs` (which is sufficiently complex already) with the extra branching structure introduced here.

I'm going to open a `[DNM]` PR where I:

1. Take a snapshot of this PR.
2. Revert all changes to `optimize/peek.rs` and `inner/peek.rs` made here.
3. Propose an alternative structure that uses a dedicated `optimize/copy_to.rs`, but piggy-backs on the `inner/peek.rs` in a slightly different way.

I hope that this will satisfy both @MaterializeInc/adapter and @MaterializeInc/compute. If we reach an agreement, I suggest to cherry-pick my commits from this PR and incorporate them here.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner/peek.rs:95]: I think it's probably better to do that in the `plan_copy_to` stage and produce a `CopyToPlan` that has `uri: http::Uri`.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner/peek.rs:1077]: What other things (besides shipping the dataflow) do you intend to do here?

Is the `session` meant to block until (a) the `COPY TO` is completed or (b) do we want to do this asynchronously.

If (a) I suggest to look at `subscribe_finish` for reference. If (b) maybe `create_materialized_view_finish` is a better starting point.

File: src/adapter/src/optimize/peek.rs
Comment (aalexandrov) [inline src/adapter/src/optimize/peek.rs:388]: @mjibson can I understand that `SELECT` can happen in a `TimestampContext::NoTimestamp` context, but do you think that we should allow this for `COPY TO`. If yes - what should be the timestamp for the configured S3 sink?

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs:1077]: The session is meant to block. Yup, IMO it should look similar to a subscribe. I will do that in a follow up PR to limit the scope in the current one.

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:388]: It should be possible to write `COPY (SELECT 1) TO ...` which dumps that one row to s3. We actually intended to use that as a quick check that the aws connection + perms are set up correctly.

**Comment by moulimukherjee** [comment]
@aalexandrov Thank you for going through the changes and creating the branch 🙇‍♀️. Let me update this PR based on the changes from https://github.com/MaterializeInc/materialize/pull/24903

File: src/adapter/src/optimize/peek.rs
Comment (aalexandrov) [reply src/adapter/src/optimize/peek.rs:390]: Then maybe some of the changes in this this bit of the code that I'm proposing in `optimize/copy_to.rs` in the other PR should be reverted (return the plan and not a `Result`), don't return an `OptimizerError::Internal` when the timestamp is not present.

Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs:390]: Yup, will remove that bit 👍

File: src/adapter/src/coord.rs
Comment (moulimukherjee) [inline src/adapter/src/coord.rs:477]: Just pulled the changes from https://github.com/MaterializeInc/materialize/pull/24903.

cc @aalexandrov @mjibson Instead of using `Either` would generic traits be better here ( I will take a stab at it then)? Because it's more tightly coupled than implied here, because `peek::Optimizer` can only be associated with `peek::GlobalMirPlan`.

Comment (aalexandrov) [reply src/adapter/src/coord.rs:477]: I went with `Either` because it is the least cognitive overhead. I'm not sure that introducing traits with associated types will make this easier to maintain - at some point we might want to change something for the `peek` branch and not for the `copy_to` branch in a place where the trait-based design dictates that both branches should be defined in terms of the same generic source code. That being said I'm not opposed to the approach if somebody else wants to take a stab.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (aalexandrov) [reply src/adapter/src/coord/sequencer/inner/peek.rs:98]: > We had chatted about doing the resolution while sequencing

Ah, OK - @mjibson is the authority for this piece of the code, so if he recommended this go with that.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner/peek.rs:812]: Note that this is adding back things that are no longer in `main`:

https://github.com/MaterializeInc/materialize/blob/783400d213914ba591c765d67667c735a6568532/src/adapter/src/coord/sequencer/inner/peek.rs#L657-L673

Can you please align the changes so the `Either::Left(optimizer)` branch is identical to what is currently in `main`?

Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs:812]: ack

Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs]: I was actually behind main on my branch on an earlier commit, just pulled and rebased again. So, should be good.

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [reply src/adapter/src/coord/sequencer/inner/peek.rs:98]: I'm fine with this being wherever makes the most sense. One glance a while ago seemed like it was a bad fit inside the optimizer directory (that was many revisions ago). Things have probably changed since then and with optimizer folks now present, I don't care at all where this goes. It certainly doesn't need to be in the sequencer.

Comment (moulimukherjee) [reply src/adapter/src/coord/sequencer/inner/peek.rs:98]: In even an earlier (🙈) version (one of the commits in https://github.com/MaterializeInc/materialize/pull/24355) it was in the `plan_copy_to` in `dml.rs`. Should it be moved back there or okay to have this in the sequencer?

File: src/adapter/src/optimize/peek.rs
Comment (moulimukherjee) [reply src/adapter/src/optimize/peek.rs]: (done, and added a test)

File: src/adapter/src/coord/sequencer/inner/peek.rs
Comment (mjibson) [reply src/adapter/src/coord/sequencer/inner/peek.rs:98]: I don't have an opinion on that so leave it where it is. If we find motivation to put it in a specific place we can move it later.

**Review by mjibson — APPROVED** [review]
Changes from Alexander LGTM.

**Pull request merged**
